### PR TITLE
begin/cleanup query notifications

### PR DESCRIPTION
**Pull request by maciejgrzybek**
Per https://github.com/prestodb/presto/pull/6114#issuecomment-261607251, removed notifications for DDLs.

---

Supersedes https://github.com/prestodb/presto/pull/6114

CC: @electrum, @dain

### CODE REVIEW

**Comment by maciejgrzybek** [comment]
@electrum do you have any feedback to it?
IMO this is implemented as agreed in https://github.com/prestodb/presto/issues/6114 (improvements suggested by @dain as a separate PR at some point in future)

**Comment by maciejgrzybek** [comment]
@electrum gentle reminder...

**Comment by maciejgrzybek** [comment]
Resolved conflicts and rebased onto the most recent master.

**Comment by maciejgrzybek** [comment]
@martint, @electrum: any thoughts on this?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java [hunk 1]
```diff
@@ -99,6 +104,8 @@
     private final List<Expression> parameters;
     private final SplitSchedulerStats schedulerStats;
 
+    private Set<ConnectorId> connectors;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: Is this just to carry result state from `doAnalyzeQuery()` to `start()`. If so, I'd just add this to PlanRoot, instead.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 2]
```diff
@@ -422,6 +423,11 @@ public TableHandle getTableHandle(Table table)
         return tables.get(table);
     }
 
+    public Collection<TableHandle> getTableHandles()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I'd call this `getTables`

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java [hunk 3]
```diff
@@ -558,6 +564,31 @@ public void beginQuery(Session session, Set<ConnectorId> connectors)
...
+    {
+        catalogsByQueryId.putIfAbsent(queryId.getId(), new ArrayList<>());
+        catalogsByQueryId.get(queryId.getId()).add(metadata);
+    }
+
+    @Override
+    public void cleanupQuery(Session session)
+    {
+        try {
+            Collection<ConnectorMetadata> catalogs = catalogsByQueryId.get(session.getQueryId().getId());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java]: Why `catalogs`? Shouldn't it be `connectors`?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 4]
```diff
@@ -575,13 +581,27 @@ public boolean transitionToFinishing()
 
     private boolean transitionToFinished()
     {
+        try {
+            metadata.cleanupQuery(session);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: In the previous commit, cleanup was an async call. Why the change to synchronous?

File: presto-main/src/main/java/com/facebook/presto/testing/TestingHandle.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package com.facebook.presto.testing;
+
+import com.facebook.presto.spi.ConnectorInsertTableHandle;
+import com.facebook.presto.spi.ConnectorOutputTableHandle;
+import com.facebook.presto.spi.ConnectorTableLayoutHandle;
+
+public enum TestingHandle
+        implements
+        ConnectorOutputTableHandle,
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingHandle.java]: Place these on the same line as `implements`

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 4 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: We had a disagreement about how does it should be, please have a look: https://github.com/prestodb/presto/pull/6114#issuecomment-260452150

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java (hunk 3 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java]: When you have two catalogs of the same type and both participate in query, then you want both catalogs being notified.
E.g.:
`INSERT INTO hive_finance.reports.stats SELECT * FROM hive_hr.salaries.engineers;`
Both: `hive_hr` and `hive_finance` should be notified.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java (hunk 1 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: Fixed in 375c67798023109ca3ec1e6ed6e0453aea1203e7.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 2 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Fixed in 375c67798023109ca3ec1e6ed6e0453aea1203e7.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingHandle.java (hunk 5 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingHandle.java]: Fixed in 2e88cfb77d40f88bf2ceb78524257bada3ad30bc.

**Comment by maciejgrzybek** [comment]
Removed async version of `cleanupQuery` per offline discussion with @martint. Docs adjusted.
